<a href="https://colab.research.google.com/github/mikeninerbravog/POC_ML_AI-in-Finance/blob/master/POC_ML_FraudDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
"""
O código inclui três demonstrações didáticas:
1. **Detecção de Fraudes**: Uso do algoritmo *Isolation Forest* para identificar transações suspeitas.
2. **Previsão de Tendência de Mercado**: Modelo de *Regressão Linear* para prever preços futuros.
3. **Chatbot Financeiro**: Simulação de um assistente virtual que responde perguntas básicas sobre finanças.
"""

import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
from sklearn.linear_model import LinearRegression
import random

# 1. Simulação de Detecção de Fraudes e Transações Válidas
def detect_transactions():
    np.random.seed(42)
    transactions = np.random.normal(100, 20, 100).tolist()  # Transações normais
    transactions.extend([500, 600, 750, 1000])  # Adicionando fraudes
    df = pd.DataFrame({'transaction_value': transactions})

    model = IsolationForest(contamination=0.05)
    df['fraudulent'] = model.fit_predict(df[['transaction_value']])

    frauds = df[df['fraudulent'] == -1]
    valid_transactions = df[df['fraudulent'] == 1]

    print("Transações fraudulentas detectadas:\n", frauds)
    print("Transações válidas detectadas:\n", valid_transactions)
    return df

# 2. Previsão de Tendência de Mercado
def market_trend_prediction():
    np.random.seed(42)
    days = np.arange(1, 101)
    prices = 50 + np.cumsum(np.random.normal(0, 1, 100))  # Simulação de preços

    model = LinearRegression()
    model.fit(days.reshape(-1, 1), prices)
    future_prices = model.predict([[101], [102], [103]])

    print("Previsão para os próximos 3 dias:", future_prices)
    return prices, future_prices

# 3. Chatbot Financeiro Simples
def chatbot_finance(question):
    """
    Chatbot simples para responder perguntas sobre finanças.
    Esse chatbot pode ser expandido para um assistente inteligente e autônomo,
    capaz de fornecer respostas mais sofisticadas e interagir dinamicamente com os usuários.
    O resultado aqui é apenas um exemplo para a PoC.
    """
    responses = {
        "O que é um investimento?": "Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro.",
        "Como evitar fraudes financeiras?": "Use cartões seguros, desconfie de ofertas muito boas e monitore suas transações.",
        "O que é um mercado de ações?": "É onde pessoas compram e vendem partes de empresas chamadas ações.",
    }
    return responses.get(question, "Desculpe, não entendi sua pergunta.")

# Executando as funções
detect_transactions()
market_trend_prediction()

# Exemplo de interação com o chatbot
pergunta = "O que é um investimento?"
print(f"Usuário: {pergunta}")
print(f"Chatbot: {chatbot_finance(pergunta)}")


Transações fraudulentas detectadas:
      transaction_value  fraudulent
31          137.045564          -1
74           47.605098          -1
100         500.000000          -1
101         600.000000          -1
102         750.000000          -1
103        1000.000000          -1
Transações válidas detectadas:
     transaction_value  fraudulent
0          109.934283           1
1           97.234714           1
2          112.953771           1
3          130.460597           1
4           95.316933           1
..                ...         ...
95          70.729701           1
96         105.922406           1
97         105.221105           1
98         100.102269           1
99          95.308257           1

[98 rows x 2 columns]
Previsão para os próximos 3 dias: [37.12727563 36.9992055  36.87113536]
Usuário: O que é um investimento?
Chatbot: Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro.


Aqui está a explicação dos resultados gerados pelo código:

### **1. Detecção de Fraudes e Transações Válidas**
A função `detect_transactions()` utilizou o **Isolation Forest**, um modelo de machine learning baseado em anomalias, para classificar transações como fraudulentas ou válidas.

#### **Transações Fraudulentas Detectadas**
Essas transações foram classificadas com **fraudulent = -1**, indicando que o modelo identificou padrões suspeitos. Os valores **500, 600, 750 e 1000** foram claramente considerados fraudes porque são **muito maiores do que a média das outras transações**, o que sugere que são outliers. Além disso, duas outras transações, **137.04 e 47.60**, também foram consideradas suspeitas, possivelmente por estarem em extremos da distribuição.

#### **Transações Válidas Detectadas**
As transações restantes foram classificadas com **fraudulent = 1**, indicando que foram consideradas normais pelo modelo. Esses valores estão mais próximos da média do conjunto de dados (em torno de 100, com desvio padrão de 20), o que justifica sua classificação como legítimos.

---

### **2. Previsão de Tendência de Mercado**
A função `market_trend_prediction()` aplicou um modelo de **Regressão Linear** para prever os preços de mercado nos próximos três dias. O modelo indicou que os preços esperados são:

- **Dia 101:** 37.13
- **Dia 102:** 37.00
- **Dia 103:** 36.87

Esses valores sugerem uma **ligeira tendência de queda**, baseada nos dados históricos de preços simulados.

---

### **3. Chatbot Financeiro**
O chatbot respondeu à pergunta **"O que é um investimento?"** com a definição simples:

> "Investimento é quando você usa dinheiro para comprar algo esperando que ele valha mais no futuro."

Isso exemplifica como a IA pode ser usada para automatizar consultas financeiras comuns.



---


### **Code Breakdown in English**

The code presents three educational demonstrations:

1. **Fraud Detection**: Uses the *Isolation Forest* algorithm to identify suspicious transactions.
2. **Market Trend Prediction**: Implements a *Linear Regression* model to forecast future prices.
3. **Financial Chatbot**: Simulates a virtual assistant that answers basic financial questions.

---

### **Code Explanation**

#### **1. Fraud Detection and Valid Transactions**
The function `detect_transactions()` simulates financial transactions and applies **Isolation Forest**, an anomaly-based machine learning model, to classify them as **fraudulent (-1)** or **valid (1)**.

##### **How It Works**
- A dataset of **100 normal transactions** is generated using a **normal distribution** (mean = 100, standard deviation = 20).
- Four **outlier transactions** (**500, 600, 750, 1000**) are manually added to simulate fraud.
- The **Isolation Forest** model is trained with **5% contamination**, assuming around 5% of the data consists of anomalies.
- The model assigns a **fraudulent (-1)** or **valid (1)** label to each transaction.

##### **Results**
- **Fraudulent Transactions Detected**: Transactions **500, 600, 750, and 1000** were identified as fraudulent because they are significantly **higher than the average transaction value (100)**, making them outliers.
- Some additional values (**137.04 and 47.60**) were also flagged as fraudulent, possibly due to being at the extreme ends of the distribution.
- **Valid Transactions**: The rest of the transactions, which are closer to the normal range (mean = 100, std = 20), were classified as legitimate.

---

#### **2. Market Trend Prediction**
The function `market_trend_prediction()` uses **Linear Regression** to predict stock prices over the next three days.

##### **How It Works**
- A **synthetic dataset** of **100 days of stock prices** is created.
- Each price is computed using a **cumulative sum** of normally distributed random values (mean = 0, std = 1).
- The **Linear Regression** model is trained using the generated dataset.
- Predictions for **days 101, 102, and 103** are made.

##### **Results**
The model forecasts:
- **Day 101:** 37.13
- **Day 102:** 37.00
- **Day 103:** 36.87

This indicates a **slight downward trend**, based on historical price data.

---

#### **3. Simple Financial Chatbot**
The function `chatbot_finance(question)` is a **basic rule-based chatbot** that answers common financial questions.

##### **How It Works**
- A dictionary stores **predefined responses** for three financial questions:
  - **"What is an investment?"** → "Investment is when you use money to buy something expecting it to be worth more in the future."
  - **"How to avoid financial fraud?"** → "Use secure cards, be wary of deals that seem too good, and monitor transactions."
  - **"What is the stock market?"** → "It’s where people buy and sell shares of companies."
- If the user's question is not in the dictionary, it responds with: *"Sorry, I didn’t understand your question."*

##### **Example Interaction**
- **User:** *"What is an investment?"*
- **Chatbot Response:** *"Investment is when you use money to buy something expecting it to be worth more in the future."*

This chatbot serves as a **proof of concept (PoC)** and can be expanded into a more sophisticated AI-powered assistant.

---

### **Final Thoughts**
This code demonstrates three practical applications of AI in finance:
1. **Anomaly detection** for fraud prevention.
2. **Predictive analytics** for stock market forecasting.
3. **Conversational AI** for automating financial inquiries.

Each function serves as a **building block** for real-world financial applications.